specs.bzl

```diff
@@ -0,0 +1,230 @@
+
+#
+# Maven specification
+#
+
+def _maven_repository(url, user = None, password = None):
+    """
+    Generates the data map for a Maven repository specifier given the available information. If both a user
+    and password are given as arguments, it will include the access credentials in the repository spec. If one or
+    both are missing, it will just generate the repository url.
+
+    Args:
+        url: *Required* A string containing the repository url (ex: `"https://maven.google.com/"`)
+        user: *Optional* A username for this Maven repository, if it requires authentication (ex: `"bob"`)
+        password: *Optional* A username for this Maven repository, if it requires authentication (ex: `"l0bl4w"`)
+
+    Output Schema:
+        {
+            "repo_url": String
+            "credentials: Optional Map
+                {
+                    "username": String
+                    "password": String
+                }
+        }
+    """
+    if user == None or password == None:
+        return { "repo_url": url }
```

Comment:
Let's be stricter here and error out with a useful message if only one of username and password is provided.

Reply:
I implemented this, but there's no way in the skylib test suite to write a test that expects failure (probably because there's no way to catch failures in starlark), so I don't know how to test this case.